File: c2_array/L384-m.cpp
//
// Created by LiHeng on 2020/6/28.
//
#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    vector<int> ori;
    vector<int> coy;
    Solution(vector<int>& nums) {
        ori = nums;
        coy = nums;
    }

    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return coy;
    }

    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        for(int i=0; i<ori.size(); ++i)
            my_swap(ori[i], ori[rand()%(ori.size()-i)+i]);
        return ori;
    }

    void my_swap(int& m , int& n){
        int tmp = m;
        m = n;
        n = tmp;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: python-algorithm/algo_09_math/L384-m.py
#--coding:utf-8--

class Solution:

    def __init__(self, nums: List[int]):
        self.origin = nums[:]  # 避免2个数组地址一样，在shuffle时候产生影响
        self.copy = nums

    def reset(self) -> List[int]:
        """
        Resets the array to its original configuration and return it.
        """
        self.copy = self.origin
        self.origin = self.origin[:]  # 避免地址一样，在shuffle时候被影响
        return self.origin

    def shuffle(self) -> List[int]:
        """
        Returns a random shuffling of the array.
        """
        # 洗牌算法
        for i in range(len(self.copy)):
            idx = random.randint(i, len(self.copy) - 1)
            self.copy[i], self.copy[idx] = self.copy[idx], self.copy[i]
        return self.copy

# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()